### PR TITLE
Update doc of link prediction

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*
#603 

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/source/advanced/link-prediction.rst [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+    where ``pos_score`` is the score of the positive edge. ``score_i`` is the score of the i-th edge. In total, there are $N+1$ edges, within which there is 1 positive edge and N negative edges.
+
+Selecting the Negative Sampler
+------------------------------
+GraphStorm provides a wide list of negative samplers:
+
+* **Uniform negative sampling**: Given ``N`` training edges under edge type ``(src_t, rel_t, dst_t)`` and the number of negatives set to K, uniform negative sampling randomly samples K nodes from ``dst_t`` for each training edge. It corrupts the training edge to form K negative edges by replacing its destination node with sampled negative nodes. In total, it will sample N * K negative nodes.
+
+    * ``uniform``: Uniformly sample K negative edges for each positive edge.
```
Comment (zhjwy9343) [inline docs/source/advanced/link-prediction.rst:99]: K -> \*K\* to show italic text.

File: docs/source/advanced/link-prediction.rst [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+    * ``fast_uniform``: same as ``uniform`` except that the sampled subgraphs
+    will not exclude edges with ``val_mask`` and ``test_mask``.
+
+    * ``all_etype_uniform``: same as ``uniform``, but it ensures that each
+    training edge type appears in every mini-batch.
+
+* **Local joint negative sampling**: Local uniform negative sampling samples negative edges in the same way as uniform negative sampling except that all the negative nodes are sampled from the local graph partition.
+
+    * ``localuniform``: Uniformly sample K negative edges for each positive edge.
```
Comment (zhjwy9343) [inline docs/source/advanced/link-prediction.rst:109]: ditto

File: docs/source/advanced/link-prediction.rst [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    However the negative nodes are sampled from the local graph partition
+    instead of being sampled globally.
+
+    * ``fast_localuniform``: same as ``localuniform`` except that the sampled subgraphs
+    will not exclude edges with ``val_mask`` and ``test_mask``. Please see the details in :ref:`speedup_lp_training_label`.
+
+* **Joint negative sampling**: Given ``N`` training edges under edge type ``(src_t, rel_t, dst_t)`` and the number of negatives set to K, joint negative sampling randomly samples K nodes from ``dst_t`` for every K training edges. For these K training edges, it corrupts each edge to form K negative edges by replacing its destination node with the same set of negative nodes. In total, it only needs to sample $N$ negative nodes. (We suppose N is dividable by K for simplicity.)
```
Comment (zhjwy9343) [inline docs/source/advanced/link-prediction.rst:116]: ditto K, and N. Use \* to wrap them for italic text.

File: docs/source/advanced/link-prediction.rst [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+    * ``joint``: Sample K negative nodes for every K positive edges.
```
Comment (zhjwy9343) [inline docs/source/advanced/link-prediction.rst:118]: ditto K

File: docs/source/advanced/link-prediction.rst [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+    The K positive edges will share the same set of negative nodes
+
+    * ``fast_joint``: same as ``joint`` except that the sampled subgraphs
+    will not exclude edges with ``val_mask`` and ``test_mask``.
+    Please see the details in :ref:`speedup_lp_training_label`.
+
+    * ``all_etype_joint``: same as ``joint``, but it ensures that each
+    training edge type appears in every mini-batch.
+
+* **Local joint negative sampling**: Local joint negative sampling samples negative edges in the same way as joint negative sampling except that all the negative nodes are sampled from the local graph partition.
```
Comment (zhjwy9343) [inline docs/source/advanced/link-prediction.rst:128]: duplicated to the line 107?

File: docs/source/advanced/link-prediction.rst [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+subgraph may have less neighbor nodes than expected as some of them would be
+removed. However, with a graph having hundreds of millions of edges (or more)
+and small validation and test sets, e.g., each with less than 10% edges, the
+impact is negligible.
+
+With DGL 1.0.4, ``fast_localuniform`` dataloader can speedup 2.4X over ``localuniform`` dataloader on training a 2 layer RGCN on MAG dataset on four g5.48x instances.
+
+Hard Negative sampling
+-----------------------
+GraphStorm provides support for users to define hard negative edges for a positive edge during Link Prediction Training.
```
Comment (zhjwy9343) [inline docs/source/advanced/link-prediction.rst:164]: "Training" -> "training"

File: docs/source/advanced/link-prediction.rst [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+Currently, hard negative edges are constructed by replacing the destination nodes of edges with pre-defined hard negatives.
+For example, given an edge (``src_pos``, ``dst_pos``) and its hard negative destination nodes ``hard_0`` and ``hand_1``, GraphStorm will construct two hard negative edges, i.e., (``src_pos``, ``hard_0``) and (``src_pos``, ``hand_1``).
```
Comment (zhjwy9343) [inline docs/source/advanced/link-prediction.rst:166]: "hand_1" -> "hard_1"

File: docs/source/advanced/link-prediction.rst [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+
+The hard negatives are stored as edge features of the target edge type.
+Users can provide the hard negatives for each edge type through ``train_etypes_negative_dstnode`` in the training config yaml.
+For example, the following yaml block defines the hard negatives for edge type (``src_type``,``rel_type0``,``dst_type``) as the edge feature ``negative_nid_field_0`` and the hard negatives for edge type (``src_type``,``rel_type1``,``dst_type``) as the edge feature ``negative_nid_field_1``.
```
Comment (zhjwy9343) [inline docs/source/advanced/link-prediction.rst:170]: "(``src_type``,``rel_type1``,``dst_type``)" -> "(``src_type, rel_type1, 
dst_type``)" ?

File: docs/source/advanced/link-prediction.rst [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+
+  .. code-block:: yaml
+
+    train_etypes_negative_dstnode:
+      - src_type,rel_type0,dst_type:negative_nid_field_0
+      - src_type,rel_type1,dst_type:negative_nid_field_1
+
+Users can also define the number of hard negatives to sample for each edge type during training though ``num_train_hard_negatives`` in the training config yaml.
```
Comment (zhjwy9343) [inline docs/source/advanced/link-prediction.rst:178]: ditto

File: docs/source/advanced/link-prediction.rst [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+
+- **A single string** The ``feature_col`` stores strings instead of string arrays. (When the input format is ``Parquet`` or ``CSV``) In this case, a ``separator`` must be provided to split the strings into node ids. The ``feature_col`` will be a 1D string list, for example ``["e0_hard_0;e0_hard_1", "e1_hard_1", ..., "en_hard_0;en_hard_1"]``. The string length, i.e., number of hard negatives, can vary from row to row. GraphStorm will automatically handle the case when some edges do not have enough hard negatives.
+
+GraphStorm will automatically translate the Raw Node IDs of hard negatives into Partition Node IDs in a DistDGL graph.
```
Comment (zhjwy9343) [inline docs/source/advanced/link-prediction.rst:206]: Can we add an example of the JSON object here to show how to use it?

File: docs/source/advanced/link-prediction.rst [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+        \end{array}
+        \right.
+        \end{eqnarray}
+
+    where ``G`` is the training graph.
+
+* **Contrastive Loss**: The contrastive loss compels the representations of connected nodes to be similar while forcing the representations of disconnected nodes remains dissimilar. In the implementation, we use the score computed by the score function to represent the distance between nodes. When computing the loss, we group one positive edge with the $N$ negative edges corresponding to it.The loss function is as follows:
+
+    .. math::
+        loss = -log(\dfrac{exp(pos_score)}{\sum_{1=0}^N exp(score_i)})
```
[the remark this answers is not in the document]
Comment (classicsong) [reply docs/source/advanced/link-prediction.rst:89]: "i=0"

[the remark this answers is not in the document]
Comment (classicsong) [reply docs/source/advanced/link-prediction.rst:89]: Updated

File: docs/source/advanced/link-prediction.rst (hunk 5 above)
Comment (classicsong) [reply docs/source/advanced/link-prediction.rst:128]: 107 is "Local uniform negative sampling". Updated.

File: docs/source/advanced/link-prediction.rst (hunk 10 above)
Comment (classicsong) [reply docs/source/advanced/link-prediction.rst:206]: I will rewrite the hard neg in another PR.

File: docs/source/advanced/link-prediction.rst [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+For example, the following yaml block defines the hard negatives for edge type (``src_type``,``rel_type0``,``dst_type``) as the edge feature ``negative_nid_field_0`` and the hard negatives for edge type ``(src_type,rel_type1,dst_type)`` as the edge feature ``negative_nid_field_1``.
+
+  .. code-block:: yaml
+
+    train_etypes_negative_dstnode:
+      - src_type,rel_type0,dst_type:negative_nid_field_0
+      - src_type,rel_type1,dst_type:negative_nid_field_1
+
+Users can also define the number of hard negatives to sample for each edge type during training though ``num_train_hard_negatives`` in the training config yaml.
+For example, the following yaml block defines the number of hard negatives for edge type (``src_type``,``rel_type0``,``dst_type``) is 5 and the number of hard negatives for edge type (``src_type``,``rel_type1``,``dst_type``) is 10.
```
Comment (zhjwy9343) [inline docs/source/advanced/link-prediction.rst:179]: Sorry, the comment should be in this line, i.e., 

"(``src_type``,``rel_type0``,``dst_type``)" -> "(``src_type, rel_type0, 
dst_type``)"  and
"(``src_type``,``rel_type1``,``dst_type``)" -> "(``src_type, rel_type1, 
dst_type``)"

File: docs/source/advanced/link-prediction.rst [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+
+Hard Negative sampling
+-----------------------
+GraphStorm provides support for users to define hard negative edges for a positive edge during Link Prediction training.
+Currently, hard negative edges are constructed by replacing the destination nodes of edges with pre-defined hard negatives.
+For example, given an edge (``src_pos``, ``dst_pos``) and its hard negative destination nodes ``hard_0`` and ``hard_1``, GraphStorm will construct two hard negative edges, i.e., (``src_pos``, ``hard_0``) and (``src_pos``, ``hard_1``).
+
+The hard negatives are stored as edge features of the target edge type.
+Users can provide the hard negatives for each edge type through ``train_etypes_negative_dstnode`` in the training config yaml.
+For example, the following yaml block defines the hard negatives for edge type (``src_type``,``rel_type0``,``dst_type``) as the edge feature ``negative_nid_field_0`` and the hard negatives for edge type ``(src_type,rel_type1,dst_type)`` as the edge feature ``negative_nid_field_1``.
```
Comment (zhjwy9343) [inline docs/source/advanced/link-prediction.rst:170]: "(``src_type``,``rel_type0``,``dst_type``)" -> "(``src_type, rel_type0, 
dst_type``)"

File: docs/source/advanced/link-prediction.rst [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+        \end{eqnarray}
+
+    where ``G`` is the training graph.
+
+* **Contrastive Loss**: The contrastive loss compels the representations of connected nodes to be similar while forcing the representations of disconnected nodes remains dissimilar. In the implementation, we use the score computed by the score function to represent the distance between nodes. When computing the loss, we group one positive edge with the ``N`` negative edges corresponding to it.The loss function is as follows:
+
+    .. math::
+        loss = -log(\dfrac{exp(pos_score)}{\sum_{i=0}^N exp(score\_i)})
+
+    where ``pos_score`` is the score of the positive edge. ``score_i`` is the score of the i-th edge. In total, there are $N+1$ edges, within which there is 1 positive edge and ``N`` negative edges.
```
Comment (zhjwy9343) [inline docs/source/advanced/link-prediction.rst:91]: "\$N+1\$" -> "\``N+1\``"

File: docs/source/advanced/link-prediction.rst [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+    where ``y`` is 1 when ``e`` is a positive edge and 0 when it is a negative edge. ``score`` is the score value of ``e`` computed by the score function.
+
+* **Weighted Cross Entropy Loss**: The weighted cross entropy loss is similar to **Cross Entropy Loss** except that it allows users to set a weight for each positive edge. The loss function of an edge ``e`` is as:
+
+    .. math::
+        \begin{eqnarray}
+            loss = - w_e \left[ y \cdot \log score + (1 - y) \cdot \log (1 - score) \right]
+        \end{eqnarray}
+
+    where ``y`` is 1 when ``e`` is a positive edge and 0 when it is a negative edge. ``score``is the score value of ``e`` computed by the score function, ``w_e`` is the weight of ``e`` and is defined as
```
Comment (zhjwy9343) [inline docs/source/advanced/link-prediction.rst:72]: Add a white space between the "\``score\``" and "is".

File: docs/source/advanced/link-prediction.rst [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+            loss = - y \cdot \log score + (1 - y) \cdot \log (1 - score)
+        \end{eqnarray}
+
+    where ``y`` is 1 when ``e`` is a positive edge and 0 when it is a negative edge. ``score`` is the score value of ``e`` computed by the score function.
+
+* **Weighted Cross Entropy Loss**: The weighted cross entropy loss is similar to **Cross Entropy Loss** except that it allows users to set a weight for each positive edge. The loss function of an edge ``e`` is as:
+
+    .. math::
+        \begin{eqnarray}
+            loss = - w_e \left[ y \cdot \log score + (1 - y) \cdot \log (1 - score) \right]
```
Comment (zhjwy9343) [inline docs/source/advanced/link-prediction.rst:69]:
```suggestion
            loss = - w\_e \left[ y \cdot \log score + (1 - y) \cdot \log (1 - score) \right]
```

File: docs/source/advanced/link-prediction.rst [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+        \end{eqnarray}
+
+    where ``y`` is 1 when ``e`` is a positive edge and 0 when it is a negative edge. ``score``is the score value of ``e`` computed by the score function, ``w_e`` is the weight of ``e`` and is defined as
+
+    .. math::
+        \begin{eqnarray}
+        w_e = \left \{
```
Comment (zhjwy9343) [inline docs/source/advanced/link-prediction.rst:76]:
```suggestion
        w\_e = \left \{
```

File: docs/source/advanced/link-prediction.rst [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+        \end{array}
+        \right.
+        \end{eqnarray}
+
+    where ``G`` is the training graph.
+
+* **Contrastive Loss**: The contrastive loss compels the representations of connected nodes to be similar while forcing the representations of disconnected nodes remains dissimilar. In the implementation, we use the score computed by the score function to represent the distance between nodes. When computing the loss, we group one positive edge with the ``N`` negative edges corresponding to it.The loss function is as follows:
+
+    .. math::
+        loss = -log(\dfrac{exp(pos_score)}{\sum_{i=0}^N exp(score\_i)})
```
Comment (zhjwy9343) [inline docs/source/advanced/link-prediction.rst:89]:
```suggestion
        loss = -log(\dfrac{exp(pos\_score)}{\sum_{i=0}^N exp(score\_i)})
```

File: docs/source/advanced/link-prediction.rst (hunk 14 above)
Comment (zhjwy9343) [inline docs/source/advanced/link-prediction.rst:91]:
```suggestion
    where ``pos_score`` is the score of the positive edge. ``score_i`` is the score of the i-th edge. In total, there are ``N+1`` edges, within which there is 1 positive edge and ``N`` negative edges.
```

File: docs/source/advanced/link-prediction.rst (hunk 13 above)
Comment (zhjwy9343) [inline docs/source/advanced/link-prediction.rst:170]:
```suggestion
For example, the following yaml block defines the hard negatives for edge type (``src_type, rel_type0, dst_type``) as the edge feature ``negative_nid_field_0`` and the hard negatives for edge type ``(src_type, rel_type1, dst_type)`` as the edge feature ``negative_nid_field_1``.
```

File: docs/source/advanced/link-prediction.rst (hunk 12 above)
Comment (zhjwy9343) [inline docs/source/advanced/link-prediction.rst:179]:
```suggestion
For example, the following yaml block defines the number of hard negatives for edge type (``src_type, rel_type0, dst_type``) is 5 and the number of hard negatives for edge type (``src_type, rel_type1, dst_type``) is 10.
```

File: docs/source/advanced/link-prediction.rst (hunk 15 above)
Comment (zhjwy9343) [inline docs/source/advanced/link-prediction.rst:72]:
```suggestion
    where ``y`` is 1 when ``e`` is a positive edge and 0 when it is a negative edge. ``score`` is the score value of ``e`` computed by the score function, ``w_e`` is the weight of ``e`` and is defined as
```

**Review by zhjwy9343 — APPROVED** [review]
LGTM

**Pull request merged**